src/transform-sdk/js/main.cc

```diff
@@ -0,0 +1,362 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "js_vm.h"
+
+#include <redpanda/transform_sdk.h>
+
+#include <cstdlib>
+#include <cstring>
+#include <expected>
+#include <memory>
+#include <optional>
+#include <print>
+#include <quickjs.h>
+#include <utility>
+#include <variant>
+#include <vector>
+
+namespace redpanda::js {
+
+extern "C" {
+
+// The following functions are how the user's code is injected into the Wasm
+// binary. We leave this imported functions, and RPK will inject these symbols
+// in `rpk transform build` after esbuild runs.
+
+#ifdef __wasi__
+
+#define WASM_IMPORT(mod, name)                                                 \
+    __attribute__((import_module(#mod), import_name(#name)))
+
+WASM_IMPORT(redpanda_js_provider, file_length)
+uint32_t redpanda_js_source_file_length();
+
+WASM_IMPORT(redpanda_js_provider, get_file)
+void redpanda_js_source_get_file(char* dst);
+
+#else
+
+constexpr std::string_view test_source_file = R"(
+import {onRecordWritten} from "@redpanda-data/transform-sdk";
+
+onRecordWritten((event, writer) => {
+  return writer.write(event.record);
+});
+)";
+
+uint32_t redpanda_js_source_file_length() { return test_source_file.size(); }
+
+void redpanda_js_source_get_file(char* dst) {
+    std::memcpy(dst, test_source_file.data(), test_source_file.size());
+}
+
+#endif
+}
+
+/**
+ * A custom JS class holding opaque bytes, easily convertable into common JS
+ * types.
+ */
+class record_data {
+public:
+    explicit record_data(bytes_view data)
+      : _data(data) {}
+    record_data(const record_data&) = delete;
+    record_data& operator=(const record_data&) = delete;
+    record_data(record_data&&) = default;
+    record_data& operator=(record_data&&) = default;
+
+    std::expected<qjs::value, qjs::exception>
+    text(JSContext* ctx, std::span<qjs::value> /*params*/) {
+        return qjs::value::string(ctx, std::string_view{_data});
+    }
+
+    std::expected<qjs::value, qjs::exception>
+    json(JSContext* ctx, std::span<qjs::value> /*params*/) {
+        // TODO(rockwood): This is going to be the most common case, this needs
+        // to be zero copy.
+        std::string str;
+        str.append_range(_data);
+        return qjs::value::parse_json(ctx, str);
+    }
+
+    std::expected<qjs::value, qjs::exception>
```

Comment:
suggestion (non-blocking): looks like this function could return `value`

Reply:
It has to return std::expected because the framework for building custom functions and exposing them to JS. This is wrapped in `make_record_data_class` which requires this signature.

Reply:
ah, i see. figured it was something like that but didn't connect the dots.